src/entities/swapRefactorDraft.ts

```diff
@@ -0,0 +1,376 @@
+                { blockNumber: block },
+            );
+
+            amount =
+                this.swapKind === SwapKind.GivenIn
+                    ? TokenAmount.fromRawAmount(this.outputAmount.token, result)
+                    : TokenAmount.fromRawAmount(this.inputAmount.token, result);
+        }
+
+        return amount;
+    }
+
+    private convertNativeAddressToZero(address: Address): Address {
+        return address === NATIVE_ADDRESS ? ZERO_ADDRESS : address;
+    }
+
+    public queryCallData(): string {
+        let callData: string;
+        if (this.isBatchSwap) {
+            callData = encodeFunctionData({
+                abi: balancerQueriesAbi,
+                functionName: 'queryBatchSwap',
+                args: [
+                    this.swapKind,
+                    this.swaps as BatchSwapStep[],
+                    this.assets,
+                    DEFAULT_FUND_MANAGMENT,
+                ],
+            });
+        } else {
+            callData = encodeFunctionData({
+                abi: balancerQueriesAbi,
+                functionName: 'querySwap',
+                args: [this.swaps as SingleSwap, DEFAULT_FUND_MANAGMENT],
+            });
+        }
+        return callData;
+    }
+
+    public get priceImpact(): PriceImpactAmount {
+        const paths = this.pathsImmutable;
+
+        const pathsReverse = paths.map(
+            (path) =>
+                new PathWithAmount(
+                    this.chainId,
+                    [...path.tokens].reverse(),
+                    [...path.pools].reverse(),
+                    path.outputAmount.amount, // TODO - Is this correct scale?
+                    path.inputAmount.amount,
+                ),
+        );
+
+        const amountInitial =
+            this.swapKind === SwapKind.GivenIn
+                ? this.getInputAmount(paths).amount
+                : this.getOutputAmount(paths).amount;
+
+        const amountFinal =
+            this.swapKind === SwapKind.GivenIn
+                ? this.getOutputAmount(pathsReverse).amount
+                : this.getInputAmount(pathsReverse).amount;
+
+        const priceImpact = MathSol.divDownFixed(
+            abs(amountInitial - amountFinal),
+            amountInitial * 2n,
+        );
+        return PriceImpactAmount.fromRawAmount(priceImpact);
+    }
+
+    // TODO - Need to add Limits, Call construction etc as in this PR: https://github.com/balancer/b-sdk/pull/220
+
+    // helper methods
+
+    private getSwaps(paths: PathWithAmount[]) {
+        let swaps: BatchSwapStep[] | SingleSwap;
+        if (this.isBatchSwap) {
+            swaps = [] as BatchSwapStep[];
+            if (this.swapKind === SwapKind.GivenIn) {
+                paths.map((p) => {
+                    p.pools.map((pool, i) => {
+                        (swaps as BatchSwapStep[]).push({
+                            poolId: pool, // TODO - Expect ID but currently using Address
```

Comment:
As I said above, I believe we should expect ids from the api and use it like that.